File: week3/whiteboarding/merge-intervals.py
#
# @lc app=leetcode id=56 lang=python3
#
# [56] Merge Intervals
#

# @lc code=start
from typing import List
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        res = []
        for a, b in sorted(intervals, key=lambda x: x[0]):
            if not res or res[-1][-1] < a:
                res.append([a, b])
            elif res[-1][-1] >= a and res[-1][-1] < b:
                res[-1][-1] = b
        return res
# @lc code=end

Sol = Solution()
intervals = [[1,3],[2,6],[8,10],[15,18]]
intervals = [[1,4],[4,5]]
intervals = [[1,4],[2,3]]
ans = Sol.merge(intervals)
print(ans)

File: week3/whiteboarding/mergeIntervals.js
// Leet # 56: https://leetcode.com/problems/merge-intervals/

const mergeInterval = (intervals) => {
	// if the intervals array doesn't contain any interval or only has one, return it
	if(intervals.length < 2) return intervals;

	// sort the intervals but the first value in each interval
	intervals.sort((a,b) => a[0] - b[0])

	// loop through the sorted array initializing the pointer at index 1
	for (let i = 1; i< intervals.length; i++) {

	    let current = intervals[i];
	    // create another pointer at the previous interval
	    let previous = intervals[i-1];

	    // compare the first index of the current interval is less than or equal to the second index of the previous interval, you have an overlap
	    if (current[0] <= previous[1]) {
		// change the current interval to reflect the merged interval
		intervals[i] = [previous[0], Math.max(previous[1],current[1])]

		// use the splice method to cut out the previous interval
		// Reminder: splice(start, deleteCount)
		intervals.splice(i-1, 1);

		// since the array has been shortened, decrement i to avoid skipping an interval with automatic incrementation
		i -= 1;
	    }
	}
	return intervals
};

    // time complexity --> O(n log n)

    // space complexity --> O(1)
